### ISSUE TITLE
Duplicate special tokens when loading dictionary from file

### ISSUE THREAD

**Issue by louismartin** [pytorch/fairseq#1309]
https://github.com/pytorch/fairseq/blob/eb68afca0208a040d4e91eceae86f5f22ca24b04/fairseq/data/dictionary.py#L178

When loading `dict.txt` that already contains special tokens such as `<s>` or `<pad>` (which are added by default in sentencepiece), these tokens appear twice in the fairseq dictionary.
They are added once in `Dictionary.__init__()` and a second time from the `dict.txt` file in `Dictionary.add_from_file()`.
These causes weird behaviours e.g. when using the model in https://github.com/huggingface/transformers.

Ideally `Dictionary` would not add the special tokens manually when loading an external `dict.txt` that already contains them (such as in https://github.com/huggingface/transformers). 
But I am afraid that this can break backward compatibility for people who already trained models with this "duplicated special tokens bug".

**Comment by louismartin** [comment]
Cross-referencing related bugs in HuggingFace Transformers: [https://github.com/huggingface/transformers/pull/2065](https://github.com/huggingface/transformers/pull/2065)

**Comment by myleott** [comment]
I think we should go a step further and remove all implicit special tokens, and only use explicit special tokens.

One nice way to handle backward compatibility is to add a header line to new `dict.txt` files indicating the version. Under the new version all special tokens are explicit, but if there is not header then we fall back to the old logic (where extra special tokens are added implicitly).

What do you think @louismartin?

**Comment by louismartin** [comment]
Yes that would definitely solve it.
Not sure there's a better way to handle backward compatibility.
Thanks!

**Comment by Junpliu** [comment]
So sorry to interrupt. At the bpe process of pretraining, is it right that fariseq did not do special preprocessing to special tokens like "<mask>" or "<sep>" ?  For example, the token ”<mask>“ in "A has done <mask> to B" would be viewed as something like separate "<m a s k >" by bpe. 
I am not sure whether I think it right. Thank you!
